pkg/virt-handler/virtwrap/cache/cache.go

```diff
@@ -187,71 +187,47 @@ func NewSharedInformer(c virtwrap.Connection) (cache.SharedInformer, error) {
 func callback(d virtwrap.VirDomain, event *libvirt.DomainEventLifecycle, watcher chan watch.Event) {
 	domain, err := NewDomain(d)
 	if err != nil {
-		logging.DefaultLogger().Error().Reason(err).Msg("Could not create the Domain.")
+		logging.DefaultLogger().Error().Object(domain).Reason(err).Msg("Could not create the Domain.")
 		watcher <- watch.Event{Type: watch.Error, Object: &v1.Status{Status: v1.StatusFailure, Message: err.Error()}}
 		return
 	}
-	logging.DefaultLogger().Info().Msgf("event received: %v:%v", event.Event, event.Detail)
+	logging.DefaultLogger().Info().Object(domain).Msgf("event received: %v:%v", event.Event, event.Detail)
 	// TODO In case of other events, it might not be enough to just send state and domainxml, maybe we have to embed the event and the details too
 	//      Think about device removal: First event is a DEFINED/UPDATED event and then we get the REMOVED event when it is done (is it that way?)
-	switch event.Event {
 
-	case libvirt.DOMAIN_EVENT_STOPPED,
-		libvirt.DOMAIN_EVENT_SHUTDOWN,
-		libvirt.DOMAIN_EVENT_CRASHED,
-		libvirt.DOMAIN_EVENT_UNDEFINED:
-		// We can't count on a domain xml in these cases, but let's try it
-		if event.Event != libvirt.DOMAIN_EVENT_UNDEFINED {
-			spec, err := NewDomainSpec(d)
-			if err != nil {
-
-				if err.(libvirt.Error).Code != libvirt.ERR_NO_DOMAIN {
-					logging.DefaultLogger().Error().Reason(err).Msg("Could not fetch the Domain specification.")
-					watcher <- watch.Event{Type: watch.Error, Object: &v1.Status{Status: v1.StatusFailure, Message: err.Error()}}
-					return
-				}
-			} else {
-				domain.Spec = *spec
-			}
-		}
-		status, reason, err := d.GetState()
-		if err != nil {
+	// No matter which event, try to fetch the domain xml and the state. If we get a IsNotFound error, that means that the VM was removed.
+	spec, err := NewDomainSpec(d)
+	if err != nil {
 
-			if err.(libvirt.Error).Code != libvirt.ERR_NO_DOMAIN {
-				logging.DefaultLogger().Error().Reason(err).Msg("Could not fetch the Domain state.")
-				watcher <- watch.Event{Type: watch.Error, Object: &v1.Status{Status: v1.StatusFailure, Message: err.Error()}}
-				return
-			}
-			domain.SetState(api.NoState, api.ReasonUnknown)
-		} else {
-			domain.SetState(convState(status), convReason(status, reason))
-		}
-	default:
-		spec, err := NewDomainSpec(d)
-		if err != nil {
-			logging.DefaultLogger().Error().Reason(err).Msg("Could not fetch the Domain specification.")
+		if !virtwrap.IsNotFound(err) {
+			logging.DefaultLogger().Error().Object(domain).Reason(err).Msg("Could not fetch the Domain specification.")
```

Comment:
When you're at it, this could be moved to some small helper function since the same code is used below.